src/main/scala/io/picnicml/doddlemodel/cluster/DBSCAN.scala

```diff
@@ -0,0 +1,87 @@
+package io.picnicml.doddlemodel.cluster
+
+import breeze.linalg.functions.euclideanDistance
+import cats.syntax.option._
+import io.picnicml.doddlemodel.data.Features
+import io.picnicml.doddlemodel.typeclasses.Clusterer
+
+/** An immutable DBSCAN model.
+  *
+  * @param eps: the maximum distance between points in a group
+  * @param minSamples: the minimum number of point in a core group
+  *
+  * Examples:
+  * val model = DBSCAN()
+  * val model = DBSCAN(eps = 1.5)
+  * val model = DBSCAN(minSamples = 3)
+  * val model = DBSCAN(eps = 2.0, minSamples = 3)
+  */
+case class DBSCAN private(eps: Double, minSamples: Int, private val label: Option[Array[Int]])
+
+object DBSCAN {
+
+  val NOISE: Int = -1
+  val UNASSIGNED: Int = Int.MaxValue
+
+  def apply(eps: Double = 1.0, minSamples: Int = 1): DBSCAN = {
+    require(eps > 0.0, "Maximum distance needs to be larger than 0")
+    require(minSamples > 0, "Minimum number of samples needs to be larger than 0")
+    DBSCAN(eps, minSamples, none)
+  }
+
+  implicit lazy val ev: Clusterer[DBSCAN] = new Clusterer[DBSCAN] {
+
+    override def isFitted(model: DBSCAN): Boolean = model.label.isDefined
+
+    override protected def labelSafe(model: DBSCAN): Array[Int] = model.label.get
+
+    override protected def copy(model: DBSCAN): DBSCAN =
+      model.copy()
+
+    override protected def copy(model: DBSCAN, label: Array[Int]): DBSCAN =
+      model.copy(label = label.some)
+
+    override protected def fitSafe(model: DBSCAN, x: Features): DBSCAN = {
+      distanceMap.clear
+      val label = Array.fill[Int](x.rows)(UNASSIGNED)
+      var groupId = 0
+      for (pointId <- 0 until x.rows if label(pointId) == UNASSIGNED) {
+        var groupQueue = findNeighbors(pointId, x, model.eps)
+        if (groupQueue.size + 1 < model.minSamples) {
+          label(pointId) = NOISE
+        } else {
+          label(pointId) = groupId
+          while (groupQueue.nonEmpty) {
+            val tmpGroupQueue = groupQueue
+            groupQueue = Set[Int]()
+            tmpGroupQueue.foreach { i =>
+              if (label(i) == NOISE) label(i) = groupId
+              else if (label(i) == UNASSIGNED) {
+                label(i) = groupId
+                val neighbors = findNeighbors(i, x, model.eps)
+                if (neighbors.size + 1 >= model.minSamples)
+                  groupQueue ++= neighbors
+              }
+            }
+          }
+          groupId += 1
+        }
+      }
+      copy(model, label)
+    }
+
+    override protected def fitPredictSafe(model: DBSCAN, x: Features): Array[Int] =
+      labelSafe(fitSafe(model, x))
+
+    private val distanceMap = scala.collection.mutable.Map[(Int, Int), Double]()
```

Comment:
I think that all the distances are needed at some point in the algorithm, so you could precompute all of them (well, 1 distance per pair, as you are doing now) in the `fitSafe(...)` method and pass it around when needed. 

That way you wouldn't need to clear the map on every call of `fitSafe(...)` and you could just as well use an immutable `Map` (you can create a `List` of tuples and do a `.toMap` at the end). Maybe the compiler might even make some optimizations regarding memory organization (I'm not 100% sure, don't quote me on that one :smile: ). 

One downside that this would bring is that `findNeighbors` would be a bit uglier.

Reply:
What is the point of doing this change?

Reply:
A suggestion for removing immutable constructs (I mentioned I'd let you know if I think of something)

Reply:
I agree with @matejklemen, no state should be kept in evidence `ev` which could get imported in some completely different part of the application.

Reply:
That make sense.
I will do the modification.